test/symforce_rust_codegen_test.py

```diff
@@ -0,0 +1,55 @@
+# ----------------------------------------------------------------------------
+# SymForce - Copyright 2022, Skydio, Inc.
+# This source code is under the Apache 2.0 license found in the LICENSE file.
+# ----------------------------------------------------------------------------
+
+import symforce
+
+symforce.set_epsilon_to_symbol()
+
+import itertools
+import functools
+
+import symforce.symbolic as sf
+from symforce import path_util
+from symforce import typing as T
+from symforce.codegen import Codegen
+from symforce.codegen import CodeGenerationException
+from symforce.codegen.backends.rust import RustConfig
+from symforce.codegen.backends.rust import ScalarType
+from symforce.test_util import TestCase
+from symforce.test_util.backend_coverage_expressions import backend_test_function
+
+TEST_DATA_DIR = (
+    path_util.symforce_data_root()
+    / "test"
+    / "symforce_function_codegen_test_data"
+    / symforce.get_symbolic_api()
+    / "symforce_rust_codegen_test"
+    / "src"
+)
```

Comment:
I'd like to actually compile the rust code to make sure we are generating valid rust code? One step further would be to write unit tests for the generated rust code. @aaron-skydio could you help me figure out the best path forward on it.

Reply:
Cool, yeah that seems like a great idea.  Some thoughts:

- I'm guessing what we want to use to build the rust code is cargo, or rustc, called directly and not managed by cmake.  In that case, I'm guessing the easiest way to do this is to write a python unit test which e.g. calls cargo to build the generated code and build a unit test that exercises the generated code, and then invokes the unit test.  Unless it's easier to make this a cmake-native test than I'm expecting; it seems like it's probably unnecessary to add rust to our cmake build just for this one test
- I think we'll want to run that python test only if a rust toolchain is available, this is similar to what we do for numba and pytorch
- For actually running that test in CI, I think we should install a rust toolchain in one or more of the jobs [here](https://github.com/symforce-org/symforce/blob/main/.github/workflows/ci.yml), and then your python unit test should run in those jobs

Does that all make sense?

Reply:
Yup! I just added it to this test. For now I think if we just build these generated files thats probably sufficient. I **think** I added rust to the ci.yml correctly but I'm not sure how to kick off ci?

Reply:
Running CI on a PR here requires a maintainer approving each run - I can do that (I'll hit the button again now), but of course that doesn't work great for you to iterate on things.  If you have GitHub Actions enabled on your fork, you can run the CI job manually on your branch from the Actions tab on your fork, like this:

![image](https://github.com/user-attachments/assets/38056c6d-b718-46c2-bb70-b1f55d9c9ab8)

Reply:
Ahh thanks. I don't use github all that much. I'm still getting quite a few errors on the cuda codegen because I changed the order in which some expressions get generated. I thought I updated all the necessary files but it still is giving me issues. Couldn't figure out tonight, but hopefully will track it down this weekend.

https://github.com/matte1/symforce/actions/runs/11809269710/job/32899285933

Reply:
Cool, looks like we're close - we run a copy of each test for each symbolic api, that's the `sympy_` or `symengine_` prefix you see on the test names there.  It looks like you've updated the symengine ones but not the sympy ones; if you run the suggested test names there it'll update the code for sympy as well, e.g. with `ctest sympy_symforce_cuda_codegen_test --update` from the build directory

Reply:
Ahh thanks! I'm now getting that the linter is failing. I fixed the mypy issues in the rust side but it seems like there are a number of linter issues in other portions of the code. Just wanted to double check thats the case and I'm not missing something on my side.

Reply:
So I see these from pylint, [here](https://github.com/matte1/symforce/actions/runs/11831654066/job/32967162488), is this what you're referring to?  I also kicked off CI again here on the PR

```
/usr/bin/python3.9 -m pylint symforce test/*.py
************* Module symforce.codegen.backends.rust.rust_config
symforce/codegen/backends/rust/rust_config.py:8:0: W0611: Unused import sympy (unused-import)
************* Module symforce.codegen.backends.rust.rust_code_printer
symforce/codegen/backends/rust/rust_code_printer.py:51:4: W0246: Useless parent or super() delegation in method '_print' (useless-parent-delegation)
symforce/codegen/backends/rust/rust_code_printer.py:107:19: W1309: Using an f-string that does not have any interpolated variables (f-string-without-interpolation)
symforce/codegen/backends/rust/rust_code_printer.py:109:19: W1309: Using an f-string that does not have any interpolated variables (f-string-without-interpolation)
symforce/codegen/backends/rust/rust_code_printer.py:11:0: W0611: Unused real imported from sympy.codegen.ast (unused-import)
symforce/codegen/backends/rust/rust_code_printer.py:12:0: W0611: Unused Rational imported from sympy.core.numbers (unused-import)
symforce/codegen/backends/rust/rust_code_printer.py:13:0: W0611: Unused S imported from sympy.core.singleton (unused-import)
************* Module symforce_rust_codegen_test
test/symforce_rust_codegen_test.py:102:17: W1309: Using an f-string that does not have any interpolated variables (f-string-without-interpolation)
test/symforce_rust_codegen_test.py:134:17: W1510: 'subprocess.run' used without explicitly defining the value for 'check'. (subprocess-run-check)
test/symforce_rust_codegen_test.py:18:0: W0611: Unused typing imported from symforce as T (unused-import)
test/symforce_rust_codegen_test.py:20:0: W0611: Unused CodeGenerationException imported from symforce.codegen (unused-import)
```

These do look like legit lint failures, let me know if any of those don't make sense (or if you were seeing different errors somewhere else)

Reply:
Yup! Your totally right, sorry I thought I got them all. Still working through some more!

Reply:
Looks like we have passing tests!
https://github.com/matte1/symforce/actions/runs/11839918889/job/32992525216

Reply:
Yay!  I'll try and get this merged soon

Reply:
Thanks for all the help with this @aaron-skydio !

Reply:
Let me know if there is anything else I can do to help get this over the line!